include/zephyr/drivers/rtc.h

```diff
@@ -0,0 +1,454 @@
+ *
+ * @param dev Device instance
+ * @param mask Mask of fields in the alarm time which are supported
+ *
+ * @return 0 if successful
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_supported_fields(const struct device *dev, uint16_t *mask);
+
+static inline int z_impl_rtc_alarm_get_supported_fields(const struct device *dev, uint16_t *mask)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_supported_fields(dev, mask);
+}
+
+/**
+ * @brief API for setting RTC alarm time
+ *
+ * @details To enable an RTC alarm, one or more fields of the RTC alarm time
+ * must be enabled. The mask designates which fields of the RTC alarm time to
+ * enable. If the mask parameter is 0, the alarm will be disabled. The RTC
+ * alarm will trigger when all enabled fields of the alarm time match the RTC
+ * time.
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Mask of fields in the alarm time to enable
+ * @param timeptr The alarm time to set
+ *
+ * @note The timeptr param may be NULL if the mask param is 0
+ *
+ * @note Only the enabled fields in the timeptr param need to be set
+ *
+ * @note Bits in the mask param are defined in the RTC_ALARM_TIME_MASK defines
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range or time is invalid
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+				 const struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_set_time(const struct device *dev, uint16_t id, uint16_t mask,
+					    const struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_set_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for getting RTC alarm time
+ *
+ * @param dev Device instance
+ * @param id Id of the alarm
+ * @param mask Destination for mask of fields which are enabled in the alarm time
+ * @param timeptr Destination for the alarm time
+ *
+ * @note bits in mask are stored in RTC_ALARM_TIME_MASK defines
+ *
+ * @return 0 if successful
+ * @return -EINVAL if id is out of range
+ * @return -ENOSYS if API is not implemented
+ * @return -ENOTSUP if API is not supported by hardware
+ * @return -errno code if failure
+ */
+__syscall int rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				 struct tm *timeptr);
+
+static inline int z_impl_rtc_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+					    struct tm *timeptr)
+{
+	const struct rtc_driver_api *api =
+		(const struct rtc_driver_api *)dev->api;
+
+	return api->alarm_get_time(dev, id, mask, timeptr);
+}
+
+/**
+ * @brief API for testing if RTC alarm is pending
```

Comment:
Is this for a pending alarm, or a pending alarm IRQ.  They are not the same.  The RTC may or may not have a connected irq line to generate interrupts.  If it doesn't an irq, it's still possible to read the current value of the event status register to see if any alarms have occurred.

Reply:
It is for clearing the pending alarm state. the functions that deal with IRQs have _irq_ in them to clearly identify them. Do you see any benefit to being able to clear the pending IRQ? I can add a rule to the API function to ensure the pending IRQ is cleared when the alarm pending state is cleared.

Reply:
No, but it wasn't clear from the brief description how it would work.  Generally this means each driver author will make up a different interpretation.

I think IRQ handling is best done in the driver.  Interaction between different alarms, and different functions on the same interrupt line like a temperature sensor, as well as level vs edge triggering, maybe needing to disable the irq in the hard irq handler context and re-enable it after masking the alarm in the rtc, a need for a driver work queue to handle i2c/spi access, etc. really needs to be in the driver.

Reply:
I like your suggestion to abstract away from the context within which the alarm callback takes place.

This will entail removing both IRQ enable and disable API calls, and renaming the `rtc_alarm_set_irq_callback` to `rtc_alarm_set_callback`. In this case, if the callback is set, alarms triggering will call back and clear pending, af callback is not set, alarms would keep the pending flag.